includes/controller/angeltypes_controller.php

```diff
@@ -340,22 +351,51 @@ function angeltypes_list_controller()
 }
 
 /**
- * Loads an angeltype from given angeltype_id request param.
+ * Validates a name for angeltypes.
+ * Returns ValidationResult containing validation success and validated name.
  *
- * @return array
+ * @param string    $name Wanted name for the angeltype
+ * @param AngelType $angeltype The angeltype the name is for
+ *
+ * @return ValidationResult result and validated name
  */
-function load_angeltype()
+function AngelType_validate_name($name, AngelType $angeltype)
 {
-    $request = request();
-    if (!$request->has('angeltype_id')) {
-        throw_redirect(page_link_to('angeltypes'));
+    $name = strip_item($name);
+    if ($name == '') {
+        return new ValidationResult(false, '');
     }
-
-    $angeltype = AngelType($request->input('angeltype_id'));
-    if (empty($angeltype)) {
-        error(__('Angeltype doesn\'t exist . '));
-        throw_redirect(page_link_to('angeltypes'));
+    if ($angeltype->id) {
+        $valid = AngelType::whereName($name)
+                ->where('id', '!=', $angeltype->id)
+                ->count() == 0;
+        return new ValidationResult($valid, $name);
     }
 
-    return $angeltype;
+    $valid = AngelType::whereName($name)->count() == 0;
+    return new ValidationResult($valid, $name);
+}
+
+/**
+ * Returns all angeltypes and subscription state to each of them for given user.
+ *
+ * @param int $userId
+ * @return Collection|AngelType[]
+ */
+function AngelTypes_with_user($userId): Collection
+{
+    return AngelType::query()
+        ->select([
+            'angel_types.*',
+            'UserAngelTypes.id AS user_angeltype_id',
+            'UserAngelTypes.confirm_user_id',
+            'UserAngelTypes.supporter',
+        ])
+        ->leftJoin('UserAngelTypes', function ($join) use ($userId) {
```

Comment:
```suggestion
        ->leftJoin('UserAngelTypes', function (JoinClause $join) use ($userId) {
```

\+ remove `/** @var JoinClause $join */`